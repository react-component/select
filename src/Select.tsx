/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 *
 * New api:
 * - listHeight
 * - listItemHeight
 * - component
 *
 * Remove deprecated api:
 * - multiple
 * - tags
 * - combobox
 * - firstActiveValue
 * - dropdownMenuStyle
 * - openClassName (Not list in api)
 *
 * Update:
 * - `backfill` only support `combobox` mode
 * - `combobox` mode not support `labelInValue` since it's meaningless
 * - `getInputElement` only support `combobox` mode
 * - `onChange` return OptionData instead of ReactNode
 * - `filterOption` `onChange` `onSelect` accept OptionData instead of ReactNode
 * - `combobox` mode trigger `onChange` will get `undefined` if no `value` match in Option
 * - `combobox` mode not support `optionLabelProp`
 */

import * as React from 'react';
import { OptionsType as SelectOptionsType } from './interface';
import SelectOptionList from './OptionList';
import Option from './Option';
import OptGroup from './OptGroup';
import { convertChildrenToData as convertSelectChildrenToData } from './utils/legacyUtil';
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
  isValueDisabled as isSelectValueDisabled,
  findValueOption as findSelectValueOption,
  flattenOptions,
  fillOptionsWithMissingValue,
} from './utils/valueUtil';
import generateSelector, { SelectProps, RefSelectProps } from './generate';
import { DefaultValueType, Key, SingleType } from './interface/generator';
import warningProps from './utils/warningPropsUtil';

const RefSelect = generateSelector<SelectOptionsType>({
  prefixCls: 'rc-select',
  components: {
    optionList: SelectOptionList,
  },
  convertChildrenToData: convertSelectChildrenToData,
  flattenOptions,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
  isValueDisabled: isSelectValueDisabled,
  findValueOption: findSelectValueOption,
  warningProps,
  fillOptionsWithMissingValue,
});

export type ExportedSelectProps<
  ValueType extends DefaultValueType = DefaultValueType
> = SelectProps<SelectOptionsType, ValueType>;

/**
 * Typescript not support generic with function component,
 * we have to wrap an class component to handle this.
 */
class Select<
  VT extends Key | Key[] = Key,
  TKeyForOptions extends Key = Key
> extends React.Component<SelectProps<SelectOptionsType<TKeyForOptions, SingleType<VT>>, VT>> {
  static Option: typeof Option = Option;

  static OptGroup: typeof OptGroup = OptGroup;

  selectRef = React.createRef<RefSelectProps>();

  focus = () => {
    this.selectRef.current.focus();
  };

  blur = () => {
    this.selectRef.current.blur();
  };

  render() {
    return <RefSelect ref={this.selectRef} {...this.props} />;
  }
}

export default Select;
