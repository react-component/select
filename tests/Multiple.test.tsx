import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import React from 'react';
import Select, { Option, OptGroup } from '../src';
import focusTest from './shared/focusTest';
import blurTest from './shared/blurTest';
import hoverTest from './shared/hoverTest';
import renderTest from './shared/renderTest';
import removeSelectedTest from './shared/removeSelectedTest';
import dynamicChildrenTest from './shared/dynamicChildrenTest';
import inputFilterTest from './shared/inputFilterTest';
import {
  expectOpen,
  toggleOpen,
  selectItem,
  injectRunAllTimers,
  findSelection,
  removeSelection,
} from './utils/common';
import allowClearTest from './shared/allowClearTest';

describe('Select.Multiple', () => {
  injectRunAllTimers(jest);

  allowClearTest('multiple', ['903']);
  focusTest('multiple');
  blurTest('multiple');
  hoverTest('multiple');
  renderTest('multiple');
  removeSelectedTest('multiple');
  dynamicChildrenTest('multiple');
  inputFilterTest('multiple');

  it('tokenize input', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select
        mode="multiple"
        optionLabelProp="children"
        tokenSeparators={[',']}
        onChange={handleChange}
        onSelect={handleSelect}
      >
        <OptGroup key="group1">
          <Option value="1">One</Option>
        </OptGroup>
        <OptGroup key="group2">
          <Option value="2">Two</Option>
        </OptGroup>
      </Select>,
    );

    wrapper.find('input').simulate('change', {
      target: {
        value: 'One',
      },
    });
    expect(handleChange).not.toHaveBeenCalled();

    handleChange.mockReset();
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two,Three',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['1', '2'], expect.anything());

    handleChange.mockReset();
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['1', '2'], expect.anything());

    expect(wrapper.find('input').props().value).toBe('');
    wrapper.update();
    expectOpen(wrapper, false);
  });

  it('tokenize input when mode=tags and open=false', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select
        mode="tags"
        optionLabelProp="children"
        tokenSeparators={[',']}
        onChange={handleChange}
        onSelect={handleSelect}
        open={false}
      >
        <Option value="0">Zero</Option>
      </Select>,
    );

    wrapper.find('input').simulate('change', {
      target: {
        value: 'One',
      },
    });
    expect(handleChange).not.toHaveBeenCalled();

    handleChange.mockReset();
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two,Three,',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['One', 'Two', 'Three'], expect.anything());

    handleChange.mockReset();
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two,',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['One', 'Two', 'Three'], expect.anything());

    expect(wrapper.find('input').props().value).toBe('');
  });

  it('shouldn\'t separate words when compositing', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select
        mode="multiple"
        optionLabelProp="children"
        tokenSeparators={[',']}
        onChange={handleChange}
        onSelect={handleSelect}
      >
        <OptGroup key="group1">
          <Option value="1">One</Option>
        </OptGroup>
        <OptGroup key="group2">
          <Option value="2">Two</Option>
        </OptGroup>
      </Select>,
    );

    wrapper.find('input').simulate('change', {
      target: {
        value: 'One',
      },
    });
    expect(handleChange).not.toHaveBeenCalled();

    handleChange.mockReset();
    wrapper.find('input').simulate('compositionstart');
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two,Three',
      },
    });
    expect(handleChange).not.toHaveBeenCalled();

    handleChange.mockReset();
    wrapper.find('input').simulate('compositionend');
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two,Three',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['1', '2'], expect.anything());

    handleChange.mockReset();
    wrapper.find('input').simulate('change', {
      target: {
        value: 'One,Two',
      },
    });
    expect(handleChange).toHaveBeenCalledWith(['1', '2'], expect.anything());

    expect(wrapper.find('input').props().value).toBe('');
    wrapper.update();
    expectOpen(wrapper, false);
  });

  it('focus', () => {
    jest.useFakeTimers();
    const handleFocus = jest.fn();
    const wrapper = mount(
      <Select mode="multiple" onFocus={handleFocus}>
        <Option value="1">One</Option>
        <Option value="2">Two</Option>
      </Select>,
    );
    wrapper.find('input').simulate('focus');
    jest.runAllTimers();
    expect(handleFocus).toHaveBeenCalled();
    jest.useRealTimers();
  });

  it('OptGroup without key', () => {
    expect(() => {
      mount(
        <Select mode="multiple" defaultValue={['1']}>
          <OptGroup label="group1">
            <Option value="1">One</Option>
          </OptGroup>
          <OptGroup label="group2">
            <Option value="2">Two</Option>
          </OptGroup>
        </Select>,
      );
    }).not.toThrow();
  });

  it('allow number value', () => {
    const handleChange = jest.fn();

    const wrapper = mount(
      <Select mode="multiple" defaultValue={[1]} onChange={handleChange}>
        <Option value={1}>1</Option>
        <Option value={2} testprop={2}>
          2
        </Option>
      </Select>,
    );

    expect(findSelection(wrapper).text()).toEqual('1');

    toggleOpen(wrapper);
    selectItem(wrapper, 1);

    expect(handleChange).toHaveBeenCalledWith(
      [1, 2],
      [expect.objectContaining({ value: 1 }), expect.objectContaining({ value: 2, testprop: 2 })],
    );
  });

  it('do not open when close button click', () => {
    const wrapper = mount(
      <Select mode="multiple">
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    toggleOpen(wrapper);
    selectItem(wrapper, 0);
    selectItem(wrapper, 1);

    toggleOpen(wrapper);
    expectOpen(wrapper, false);
    removeSelection(wrapper);
    expectOpen(wrapper, false);
    expect(wrapper.find('Selector').props().values).toEqual([
      expect.objectContaining({ value: 2 }),
    ]);
  });

  it('select when item enter', () => {
    const wrapper = mount(
      <Select mode="multiple">
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    toggleOpen(wrapper);
    wrapper
      .find('div.rc-select-item-option')
      .at(1)
      .simulate('mouseMove');

    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
    expectOpen(wrapper);
    expect(wrapper.find('Selector').props().values).toEqual([
      expect.objectContaining({ value: 2 }),
    ]);
  });

  it('enter twice to cancel the selection', () => {
    const wrapper = mount(
      <Select mode="multiple">
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    toggleOpen(wrapper);
    wrapper
      .find('div.rc-select-item-option')
      .first()
      .simulate('mousemove');
    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });

    wrapper
      .find('div.rc-select-item-option')
      .first()
      .simulate('mousemove');
    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });

    expect(wrapper.find('Selector').props().values).toEqual([]);
  });

  it('do not crash when children has empty', () => {
    const wrapper = mount(
      <Select mode="multiple">
        {null}
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    toggleOpen(wrapper);
    selectItem(wrapper);

    // Do not crash
  });

  it('do not crash when value has empty string', () => {
    const wrapper = mount(
      <Select mode="multiple" value={['']}>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    expect(findSelection(wrapper).text()).toEqual('');
  });

  it('show arrow on multiple mode when explicitly set', () => {
    const wrapper = mount(
      <Select mode="multiple" value={['']}>
        <Option value={1}>1</Option>
        <Option value={2}>2</Option>
      </Select>,
    );

    expect(wrapper.find('.rc-select-arrow-icon').length).toBeFalsy();

    wrapper.setProps({
      showArrow: true,
    });
    expect(wrapper.find('.rc-select-arrow-icon').length).toBeTruthy();
  });

  it('block input when fast backspace', () => {
    jest.useFakeTimers();
    const onChange = jest.fn();

    const wrapper = mount(
      <Select
        mode="multiple"
        value={['bamboo']}
        options={[{ value: 'bamboo' }, { value: 'light' }]}
        onChange={onChange}
      />,
    );

    // First type
    wrapper.find('input').simulate('keydown', { which: KeyCode.L });
    wrapper.find('input').simulate('change', { target: { value: 'l' } });

    // Backspace
    wrapper.find('input').simulate('keydown', { which: KeyCode.BACKSPACE });
    wrapper.find('input').simulate('change', { target: { value: '' } });

    onChange.mockReset();

    wrapper.find('input').simulate('keydown', { which: KeyCode.BACKSPACE });
    expect(onChange).not.toHaveBeenCalled();

    jest.runAllTimers();
    wrapper.find('input').simulate('keydown', { which: KeyCode.BACKSPACE });
    expect(onChange).toHaveBeenCalledWith([], expect.anything());

    jest.useRealTimers();
  });

  it('show placeholder when searchValue is controlled', () => {
    const wrapper = mount(<Select mode="multiple" searchValue="light" placeholder="bamboo" />);
    expect(wrapper.find('.rc-select-selection-placeholder').length).toBeTruthy();
    toggleOpen(wrapper);
    expect(wrapper.find('.rc-select-selection-placeholder').length).toBeFalsy();
  });

  it('clear input when popup closed', () => {
    const wrapper = mount(<Select mode="multiple" options={[{ value: 'light' }]} showSearch />);
    toggleOpen(wrapper);
    wrapper.find('input').simulate('change', { target: { value: 'bamboo' } });
    expect(wrapper.find('input').props().value).toEqual('bamboo');

    // Close and open again
    toggleOpen(wrapper);
    toggleOpen(wrapper);
    expect(wrapper.find('input').props().value).toEqual('');
  });
});
