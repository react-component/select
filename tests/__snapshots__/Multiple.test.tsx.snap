// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.Multiple render not display maxTagPlaceholder if maxTagCount not reach 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
    <span
      class="rc-select-selection-placeholder"
    />
  </div>
</div>
`;

exports[`Select.Multiple render truncates tags by maxTagCount and show maxTagPlaceholder 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            One
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Two
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-rest"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
        >
          <span
            class="rc-select-selection-item-content"
          >
            <span>
              Omitted
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select.Multiple render truncates tags by maxTagCount and show maxTagPlaceholder function 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            One
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Two
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-rest"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
        >
          <span
            class="rc-select-selection-item-content"
          >
            <span>
              1 values omitted
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select.Multiple render truncates values by maxTagTextLength 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            On...
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Tw...
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
